tariff/ngeso.go

```diff
@@ -0,0 +1,145 @@
+package tariff
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/tariff/ngeso"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"golang.org/x/exp/slices"
+)
+
+type Ngeso struct {
+	mux            sync.Mutex
+	log            *util.Logger
+	regionId       string
+	regionPostcode string
+	data           api.Rates
+	updated        time.Time
+}
+
+var _ api.Tariff = (*Ngeso)(nil)
+
+func init() {
+	registry.Add("ngeso", NewNgesoFromConfig)
+}
+
+func NewNgesoFromConfig(other map[string]interface{}) (api.Tariff, error) {
+	var cc struct {
+		Region   string
+		Postcode string
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.Region != "" && cc.Postcode != "" {
+		return nil, errors.New("region and postcode cannot be defined simultaneously - pick one")
+	}
+
+	t := &Ngeso{
+		log:            util.NewLogger("ngeso"),
+		regionId:       cc.Region,
+		regionPostcode: cc.Postcode,
+	}
+
+	done := make(chan error)
+	go t.run(done)
+	err := <-done
+
+	return t, err
+}
+
+func (t *Ngeso) run(done chan error) {
+	var once sync.Once
+	client := request.NewHelper(t.log)
+	bo := newBackoff()
+
+	// Use national results by default.
+	tUri := ngeso.ConstructForecastNationalAPI()
+	var apiDecodeAsRegional bool
+
+	// If a region is available, use that.
+	// These should never be set simultaneously (see NewNgesoFromConfig), but in the rare case that they are,
+	// use the postcode as the preferred method.
+	// Regional responses are subtly different, so we set a flag to let the loop know to decode them as such. FIXME?
+	if t.regionId != "" {
+		tUri = ngeso.ConstructForecastRegionalByIdAPI(t.regionId)
+		apiDecodeAsRegional = true
+	}
+	if t.regionPostcode != "" {
+		tUri = ngeso.ConstructForecastRegionalByPostcodeAPI(t.regionPostcode)
+		apiDecodeAsRegional = true
+	}
+
+	// Data updated by ESO every half hour, but we only need data every hour to stay current.
+	for ; true; <-time.Tick(time.Hour) {
+		// FIXME Eww, this is a sloppy way of handling this. Maybe we should move abstraction of this to the ngeso package?
+		var rgnlWrapper ngeso.RegionalIntensityResult
+		var res ngeso.IntensityRates
```

Comment:
Maybe you could make `res` an `interface` and assign the target type on top. That would leave only one `GetJSON` and clearer error handling.

Reply:
Good thinking - to be honest this whole thing probably needs a refactor, but I wanted to get something out the door before my coffee shop time ran out ☕️

Probably going to move a lot of this client logic out to the 'ngeso' package and tidy it in the process